feat: introduce VotesERC20Lockable contract

adamgall opened the pull request:
Closes ENG-568

# Add VotesERC20LockableV1 Token Contract with Comprehensive Testing

This PR introduces a new lockable version of our VotesERC20 token contract with extensive test coverage. The implementation allows for controlled token transfers through a locking mechanism and whitelist system.

## Core Features

- **Token Locking**: Owner can lock/unlock token transfers globally
- **Whitelist System**: Addresses can be whitelisted to transfer tokens even when locked
- **Owner Privileges**: 
  - Can transfer tokens regardless of lock status
  - Can mint new tokens at any time
  - Manages whitelist and lock status
- **ERC165 Support**: Full interface detection for ILockableV1, IMintableV1, and other interfaces

## Technical Details

### New Contracts
- `VotesERC20LockableV1.sol`: Main implementation extending VotesERC20
- `ILockableV1.sol`: Interface defining locking and whitelist functionality
- `IMintableV1.sol`: Interface for minting capabilities

### Key Implementation Points
- Uses OpenZeppelin's ERC165Storage for interface detection
- Implements `isTransferable` modifier for transfer control
- Supports proxy initialization with flexible parameters
- Version tracking with constant VERSION = 1

### Security Features
- All privileged functions are owner-restricted
- Events emitted for lock status changes and whitelist updates
- Comprehensive checks for lock/unlock state transitions
- Safe initialization pattern with `_disableInitializers()`

## Testing

Extensive test suite covering:
- Initialization scenarios
- Locking/unlocking functionality
- Whitelist management
- Token transfers (direct and approved)
- Minting operations
- Interface support (ERC165)
- Edge cases and access control

## Security Considerations

- Only owner can lock/unlock tokens and manage whitelist
- Minting restricted to owner regardless of lock status
- All state-changing functions emit appropriate events
- Initialization can only occur once
- Transfer restrictions properly enforced through modifier

adamgall commented:
I still have some more cleanup and ERC165 support to add to this contract, let's not merge yet

edit: done

dacebt reviewed (commented):
reviewed. no additional comments. looks good to me. I'll revisit when you finish cleanup

adamgall committed:
feat: introduce VotesERC20LockableV1 contract with locking and whitelisting functionality

This commit adds the VotesERC20LockableV1 contract, which implements a lockable ERC20 token with whitelisting capabilities. Key features include:
- Locking mechanism to restrict token transfers.
- Whitelisting functionality allowing specific addresses to transfer tokens even when locked.
- Initialization function to set up the contract with ownership and initial state.
- Comprehensive test suite covering initialization, locking, whitelisting, and transfer functionalities.

This implementation enhances the flexibility and security of token management in decentralized applications.

Commit:
Supports mint(), which should be doable even when the tokens are locked

Commit:
Changed back to _beforeTokenTransfer and simplified test

Commit:
remove .only from tests

Commit:
Simplied mint tests

adamgall committed:
Merge pull request #233 from decentdao/v-voteserc20lockable-mint

Supports mint(), which should be doable even when the tokens are locked

adamgall committed:
feat: add IMintableV1 interface and integrate with VotesERC20LockableV1

This commit introduces the IMintableV1 interface, allowing the minting of tokens. The VotesERC20LockableV1 contract is updated to implement this interface, enhancing its functionality. Additionally, tests are updated to verify support for the new minting interface.

adamgall committed:
Merge pull request #234 from decentdao/v1-add-imintable-interface

feat: add IMintableV1 interface and integrate with VotesERC20LockableV1

twodam commented on contracts/VotesERC20LockableV1.sol:
```diff
@@ -0,0 +1,102 @@
+    bool public locked;
+    mapping(address => bool) public whitelisted;
+
+    event Locked(bool locked);
+    event Whitelisted(address indexed account, bool isWhitelisted);
```
move events into interface `ILockableV1`?

adamgall replied:
Good call, will do this

adamgall replied:
Moved those events, and created custom errors, into the interface

adamgall committed:
refactor: improve locking mechanism and error handling in VotesERC20LockableV1

This commit refines the locking functionality in the VotesERC20LockableV1 contract by replacing require statements with custom error handling for better clarity and gas efficiency. The events for locking and whitelisting are moved to the ILockableV1 interface, ensuring consistency across implementations. Additionally, tests are updated to reflect these changes, verifying the new error handling behavior when attempting to lock or transfer tokens while locked.

adamgall committed:
feat(paymaster): implement function validation system with ERC20/721 voting validators

This commit introduces a comprehensive function validation system for the DecentPaymaster,
replacing the simple function whitelisting with a more flexible and secure validator-based
approach. The new system enables fine-grained control over function execution while
maintaining security guarantees through interface validation and ownership checks.

Core Architectural Changes:
- Replace basic function whitelisting with a validator contract system
- Introduce IFunctionValidator interface for standardized validation
- Add ERC165 support across all contracts for interface detection
- Return light account owner from validateSmartAccount for enhanced security context
- Implement proper interface ID calculation and validation in tests

New Components:
1. Validator Interface & Base:
   - IFunctionValidator interface defining the validation contract standard
   - ERC165 support for interface detection
   - Version tracking for future upgrades

2. Voting Validators:
   - LinearERC20VotingV1ValidatorV1:
     * Validates ERC20-based voting operations
     * Checks vote types, proposal existence, voting periods
     * Prevents double voting through hasVoted checks

   - LinearERC721VotingV1ValidatorV1:
     * Validates NFT-based voting with weight support
     * Ensures token ownership at validation time
     * Validates array lengths and vote parameters
     * Supports mixed weight configurations
     * Prevents voting with zero total weight

3. Mock Contracts for Testing:
   - MockLinearERC20VotingV1
   - MockLinearERC721VotingV1
   - MockValidator for general validation testing
   - Enhanced MockERC721 with additional minting capabilities

Security Improvements:
- Validators exactly mirror all revert conditions from voting contracts
- Strict ownership verification for NFT-based voting
- Comprehensive voting period and proposal existence checks
- Interface validation for validator contracts
- Prevention of double-voting scenarios
- Weight validation for NFT-based voting
- Enhanced error reporting with custom errors

Test Coverage:
- Comprehensive ERC165 interface testing using proper interface ID calculation
- Extensive validation scenarios for both ERC20 and ERC721 voting
- Edge case testing for array lengths, vote types, and ownership
- Mock contract integration tests
- Validation failure scenarios
- Interface support verification

Breaking Changes:
1. Function Renames:
   - whitelistFunction → setFunctionValidator
   - unwhitelistFunction → removeFunctionValidator
   - isFunctionWhitelisted → hasFunctionValidator

2. Return Type Changes:
   - validateSmartAccount now returns (bool isValid, address lightAccountOwner)
   - validateUserOp includes lightAccountOwner in return tuple

3. Interface Changes:
   - New IFunctionValidator interface requirement
   - Additional ERC165 support requirements

Migration Notes:
- Existing whitelisted functions need to be migrated to use validators
- Contracts depending on validateSmartAccount need to handle new return type
- Systems using the paymaster need to implement or use existing validators

This architectural change significantly improves the security and flexibility of
the paymaster system while maintaining backward compatibility through proper
interface versioning and ERC165 support. The new validator system allows for
more complex validation scenarios while ensuring all security properties are
maintained.

adamgall committed:
Merge branch 'v1-paymaster-vuln-fixes' into v1-voteserc20lockable

adamgall committed:
Revert "feat(paymaster): implement function validation system with ERC20/721 voting validators"

This reverts commit 5e817b1588f5e9cc20ab561496d5ab137b23fdc8.
